examples/miral-shell/spinner/miregl.cpp

```diff
@@ -192,9 +193,60 @@ void MirEglApp::make_current(EGLSurface eglsurface) const
         throw std::runtime_error("Can't eglMakeCurrent");
 }
 
-void MirEglApp::swap_buffers(EGLSurface eglsurface) const
+void MirEglApp::swap_buffers(EGLSurface eglsurface, wl_surface* wayland_surface) const
 {
+    // Taken primarily from src/platforms/wayland/displayclient.cpp
+    struct FrameSync
+    {
+        explicit FrameSync(wl_surface* surface):
+            surface{surface}
+        {
+        }
+
+        void init()
+        {
+            callback = wl_surface_frame(surface);
+            static struct wl_callback_listener const frame_listener =
+                {
+                    [](void* data, auto... args)
+                    { static_cast<FrameSync*>(data)->frame_done(args...); },
+                };
+            wl_callback_add_listener(callback, &frame_listener, this);
+        }
+
+        ~FrameSync()
+        {
+            wl_callback_destroy(callback);
+        }
+
+        void frame_done(wl_callback*, uint32_t)
+        {
+            {
+                std::lock_guard lock{mutex};
+                posted = true;
+            }
+            cv.notify_one();
+        }
+
+        void wait_for_done()
+        {
+            std::unique_lock lock{mutex};
+            cv.wait_for(lock, std::chrono::milliseconds{100}, [this]{ return posted; });
+        }
+
+        wl_surface* const surface;
+
+        wl_callback* callback;
+        std::mutex mutex;
+        bool posted = false;
+        std::condition_variable cv;
+    };
+
+    auto const frame_sync = std::make_shared<FrameSync>(wayland_surface);
+    frame_sync->init();
+    eglSwapInterval(egldisplay, 0);
     eglSwapBuffers(egldisplay, eglsurface);
+    frame_sync->wait_for_done();
```

Comment:
This seems complex.

Firstly, as written, we don't need a `shared_ptr<FrameSync>` or even a `unique_ptr<FrameSync>` as a `FrameSync` will do just as well.

Secondly, we don't need two-phase construction `FrameSync::init()` could be inlined in the ctor.

Thirdly, we don't need two-phase destruction `FrameSync::wait_for_done()` could be inlined in the dtor.